### ISSUE TITLE
回归的问题

### ISSUE THREAD

**Issue by Ezra-Yu** [XiaoJiNu/yolov5-v1-chinese-comment#1]
这里* 2你知道什么意思吗？怎么感觉值域会变化

![image](https://user-images.githubusercontent.com/18586273/90219915-b5092900-de39-11ea-8965-b399698ede95.png)

**Comment by XiaoJiNu** [comment]
@Ezra-Yu 无法查看图片，可以列出代码不

**Comment by Ezra-Yu** [comment]
https://github.com/ultralytics/yolov5/issues/731
这个问题

**Comment by Ezra-Yu** [comment]
其实我还是不太懂

**Comment by Ezra-Yu** [comment]
可以加你qq不？

**Comment by XiaoJiNu** [comment]
@Ezra-Yu v5关于坐标回归的计算方式和v3不同，我根据代码得出的公式如下，但是为什么是下面这个公式，目前我也没有弄清楚
bx = {[2*sigmoid(tx)-0.5]+cx}*stride 
by = {[2*sigmoid(ty)-0.5]+cy}*stride
bw = pw*[2*sigmoid(tw)]**2
bh = ph*[2*sigmoid(th)]**2

**Comment by Ezra-Yu** [comment]
Generally, the groundtruth bbox would be matched with three grids (contains two neighbor grids), so the regression arange should be -0.5~1.5. More details refer to build_targets()

v5的人这么回答我的，我就更不懂了。

**Comment by Ezra-Yu** [comment]
elif style == 'rect4':
                g = 0.5  # offset
                '''
                # (gxy % 1. < g): 以每个格子左上角为原点,水平，垂直方向为x,y轴的坐标系。判断每个目标方框的中心坐标x',y'(即gxy中一行元素)，在它所在的这个格子的坐标系中是否满足0<x'<0.5, 0<y'<0.5。
                # (gxy > 1.): 以输入图片左上角为原点的坐标系中(实质就是图片被划分的网格坐标系，尺寸也就是当前特征图的维度)，判断每个目标方框中心坐标x,y是否大于1
                # j：长度为31的tensor,每个元素表示每个目标方框的x坐标是否满足0<x'<0.5 且 x>1，满足为true,否则为false. true表示这个目标在网格坐标系中x大于1并且在目标所在格子中处于格 子中心的左边
                # k：长度为31的tensor，每个元素表示每个目标方框的y坐标是否满足0<y'<0.5 且 y>1，满足为true,否则为false. true表示这个目标在网格坐标系中y大于1并且在目标所在格子中处于格子中心的上边
                # l,m同理，其中为true的元素表示这个目标是否满足0.5<x'<1, 0.5<y'<1且x坐标小于当前yolo层特征图的宽w-1, y坐标小于特征图的高h-1。
                # l: 长度为31的tensor,实质是x<w-1,且在目标所在格子中心的右边的目标为true，其余为false
                # m: 长度为31的tensor,实质是y<h-1,且在目标所在格子中心的下边的目标为true，其余为false
                '''
                j, k = ((gxy % 1. < g) & (gxy > 1.)).T
                l, m = ((gxy % 1. > (1 - g)) & (gxy < (gain[[2, 3]] - 1.))).T
                # 这里是将前面由j,k,l,m提取出来的目标的真实标签与原来的目标的方框真实标签堆叠起来，增加了目标标签数量
                a, t = torch.cat((a, a[j], a[k], a[l], a[m]), 0), torch.cat((t, t[j], 									  t[k], t[l], t[m]), 0)
                '''
                off = [[1, 0], [0, 1], [-1, 0], [0, -1]]
                offsets的计算是以当前yolo层的特征图网格为坐标系
                z是原始目标的偏移量，为0. 
                z[j]+off[0], 将处于格子中心左边的目标向右移动1
                z[k]+off[1], 将处于格子中心上边的目标向下移动1
                z[l]+off[2], 将处于格子中心右边的目标向左移动1
                z[m]+off[3], 将处于格子中心下边的目标向上移动1
                重点：对于每个分配了anchor的目标，将它们在格子中相对于格子中心的位置分为上、下、左、右
                     然后将这些目标按照这个相对位置的相反方向移动1，类似于目标抖动，增加了目标的数量
                     这么做的理由是什么？？？

这里好绕啊，我没看懂，这个图难画不

**Comment by XiaoJiNu** [comment]
就是每个目标在当前这个网格中
1 以竖直线将当前格子划分为左右两部分，如果目标中心在左边，则目标会向右移动1个单位，在右边格子中会产生一个新的目标；如果在右边，则向左移动1个单位，在左边格子产生一个目标。极限情况，在右边的目标无限靠近中心，则它向左移动1，新的目标距离格子中心水平坐标无限靠近-0.5. 在左边的目标无限靠近中心，则它向右移动1，新的目标距离格子中心水平坐标无限靠近1.5.所以回归的值是-0.5到1.5
2 以水平线将当前格子划分为上下两部分，如果目标中心在上边，则目标会向下移动1个单位，在下边格子中会产生一个新的目标；如果在下边，则向上移动1个单位，在上边格子产生一个目标。同理，回归值为-0.5到1.5
看下面这个图，目标中心位于当前网格的左边，所以向右移动，位于网格上边，所以向下移动，一共得到三个目标

![图片](https://user-images.githubusercontent.com/30587645/90377999-66fb5c00-e0ab-11ea-8159-6625117ff111.png)

**Comment by Ezra-Yu** [comment]
三个目标什么意思？

**Comment by liangheming** [comment]
其实这一块的代码，最本质的改进是为了增加positive sample 的个数，原yolo的anchor分配主要策略是以grid为单位，也就是说一个gt_box，在某一个尺度的下的grid map中，只会被分配一个grid中某个特定的anchor中，而在yolov5中，一个gt_box会被分配到多个(一般是三个，就是楼上所说的三目标)grid的anchor中，具体方案其实很简单，例如对于gt中心在grid中偏向左上的，那么除了gt中心所在的grid(c)，还有该grid(c)左边的grid(l)，以及该grid(c)上边的grid(u)都会有anchor与之匹配(偏右下的情况以此类推)。本质上增加了positive sample的个数，其实这种偏密集的anchor分配方式，会遇到一种 anchor confuse的情况，就是一个anchor，被多个gt匹配上的情况，yolov5 感觉没有对这块作处理。只是后处理的时候，使用multi class的情况，将一个anchor的所有分类情况都过滤出来。个人感觉anchor confuse可以做一个改进。

**Comment by liangheming** [comment]
个人基于原有代码，重新改写了yolov5，没有pretrain，s版本，直接训练，640的size ，最终map 34.2,其实相较之前的33.1高了很多，不过后面作者有改进了原有代码，将现在性能提升到了37.0，怎么说呢,作者TQL，(主要改进在于使用了pytorch1.6自带的混合精度训练，还有就是sync bn,当然作者也添加了mixup的数据增强方式[个人感觉作者mixup方法好像有点问题])。无意中看到这个项目，发现还是有一些人会去研究代码的，而不是无脑套用。非常开心，希望能一起交流

**Comment by Ezra-Yu** [comment]
应该不是合精度和sync bn，我记得上个版本没这些，网络是有36.7的，最近的提升我看下了貌似是调参的结果。作者使用的马赛克数据增强增强可以对小目标提升。

**Comment by liangheming** [comment]
> 应该不是合精度和sync bn，我记得上个版本没这些，网络是有36.7的，最近的提升我看下了貌似是调参的结果。作者使用的马赛克数据增强增强可以对小目标提升。

36.7的这一版，我倒是没有仔细看过，不过超参的配置的确和之前有一点差别，不过调参这种操作，深究了感觉有些玄了。另外还有一些细节上的调整，比如之前数据增强的仿射变换改成了透视变换(其实这个代码改进很有用，以前在mosaic之后紧接着的那个仿射变化会导致增强后的图像出现大片空白，换成透视变换明显居中了很多)。还有像iou损失部分改成了ciou等。另外好像输出的特征层也多了一层，各层损失的balance系数也有一些调整，不过我感觉在各特征层进行计算损失之前加一个可训练的scale系数可能会更有用。感觉作者更新还是非常频繁的。哎哎，主要是个人改进一个点要跑个几天，实在顶不住。另外稍微吐槽一下，yolov5的代码组织有点乱。

**Comment by XiaoJiNu** [comment]
@liangheming 非常开心您能给出这么详细的解释，我个人对某些点还是不是很透彻，由于精力和能力有限，目前不敢说对每行代码都弄明白了，不过后面我会继续注释这些代码，希望和各位大佬一起讨论，最终能完全弄清楚这个项目。

**Comment by chenyoufa** [comment]
最后到底是什么意思，我也找了好几天这个问题的答案了？

**Comment by chenyoufa** [comment]
bx = {[2sigmoid(tx)-0.5]+cx}stride  是相邻的有两个，yolo3 是一个 grid 取值0~1 yolo5 是两个所以x2 
by = {[2sigmoid(ty)-0.5]+cy}stride  同上
bw = pw[2sigmoid(tw)]**2       这个就不懂了
bh = ph[2*sigmoid(th)]**2

**Comment by liwei-1108** [comment]
请问一下大佬，YOLO.py的这一段代码怎么解读
            args = [c1, c2, *args[1:]]
            if m in [BottleneckCSP, C3]:
                args.insert(2, n)  # number of repeats
                n = 1
